_posts/2022-05-20-changelog-state-backend.md

```diff
@@ -0,0 +1,353 @@
+      <td style="padding: 5px">1s</td>
+    </tr>
+    <tr>
+      <td style="padding: 5px">Full checkpoint size *, 99%</td>
+      <td style="padding: 5px">19.6GB</td>
+      <td style="padding: 5px">25.6GB</td>
+    </tr>
+    <tr>
+      <td style="padding: 5px">Recovery time (local recovery disabled)</td>
+      <td style="padding: 5px">20-21s</td>
+      <td style="padding: 5px">35-65s (depending on the checkpoint)</td>
+    </tr>
+  </tbody>
+</table>
+
+As can be seen from the above table, checkpoint duration is reduced 10 times for 99.9% of checkpoints, while space usage increases by 30%, and recovery time increases by 66%-225%.
+
+More details about the checkpoints (Changelog Enabled / Changelog Disabled):
+
+<table border="1">
+  <thead>
+    <tr>
+      <th style="padding: 5px">Percentile</th>
+      <th style="padding: 5px">End to End Duration</th>
+      <th style="padding: 5px">Checkpointed Data Size *</th>
+      <th style="padding: 5px">Full Checkpoint Data Size *</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <td style="padding: 5px">50%</td>
+      <td style="padding: 5px">311ms / 5s</td>
+      <td style="padding: 5px">14.8MB / 3.05GB</td>
+      <td style="padding: 5px">24.2GB / 18.5GB</td>
+    </tr>
+    <tr>
+      <td style="padding: 5px">90%</td>
+      <td style="padding: 5px">664ms / 6s</td>
+      <td style="padding: 5px">23.5MB / 4.52GB</td>
+      <td style="padding: 5px">25.2GB / 19.3GB</td>
+    </tr>
+    <tr>
+      <td style="padding: 5px">99%</td>
+      <td style="padding: 5px">1s / 7s</td>
+      <td style="padding: 5px">36.6MB / 5.19GB</td>
+      <td style="padding: 5px">25.6GB / 19.6GB</td>
+    </tr>
+    <tr>
+      <td style="padding: 5px">99.9%</td>
+      <td style="padding: 5px">1s / 10s</td>
+      <td style="padding: 5px">52.8MB / 6.49GB</td>
+      <td style="padding: 5px">25.7GB / 19.8GB</td>
+    </tr>
+  </tbody>
+</table>
+
+\* Checkpointed Data Size is the size of data persisted after receiving the necessary number of checkpoint barriers, during a so-called synchronous and then asynchronous checkpoint phases. Most of the data is persisted pre-emptively (i.e. after the previous checkpoint and before the current one), and that’s why this size is much lower when the Changelog is enabled.
+<br>
+\* Full checkpoint size is the total size of all the files comprising the checkpoint, including any files reused from the previous checkpoints. Compared to a normal checkpoint, the one with a changelog is less compact, keeping all the historical values since the last materialization, and therefore consumes much more space
+
+## Window workload
+This workload used Processing Time Sliding Window. As can be seen below, checkpoints are still faster, resulting in 3 times shorter durations; but storage amplification is much higher in this case (45 times more space consumed):
+   Checkpoint Statistics for Window Workload with Changelog Enabled / Changelog Disabled
```

Comment:
an extra line is needed here before this paragraph.

(The current line connects directly with the previous sentence in my md view)